src/clj/rems/api/applications_v2.clj

```diff
@@ -0,0 +1,869 @@
+                    :optional (:field/optional field)
+                    :options (:field/options field)
+                    :maxlength (:field/max-length field)
+                    :value (:field/value field)
+                    :previous-value nil ; TODO
+                    :localizations (into {} (for [lang (set (concat (keys (:field/title field))
+                                                                    (keys (:field/placeholder field))))]
+                                              [lang {:title (get-in field [:field/title lang])
+                                                     :inputprompt (get-in field [:field/placeholder lang])}]))})
+                 (:form/fields application))}))
+
+(defn api-get-application-v1 [user-id application-id]
+  (when-let [v2 (api-get-application-v2 user-id application-id)]
+    (transform-v2-to-v1 v2 user-id)))
+
+;;; v2 API, listing all applications
+
+(defn- applications-view
+  "Projection for the current state of all applications."
+  [applications event]
+  (if-let [app-id (:application/id event)] ; old style events don't have :application/id
+    (update applications app-id application-view event)
+    applications))
+
+(defn- reset-applications-state [_]
+  {:last-processed-event-id 0
+   :applications {}})
+
+(def ^:private applications-state
+  "Stateful projection tracking the current state of all applications.
+  Will be updated periodically by `applications-update-scheduler`,
+  but an update may also be triggered with `trigger-applications-update!`
+  e.g. after committing some new events."
+  (agent (reset-applications-state nil)
+         :error-handler (fn [_agent exception]
+                          (log/error exception "Updating projection failed"))))
+
+(declare trigger-applications-update!)
+
+(defn- update-applications! [state]
+  (let [from-id (:last-processed-event-id state)
+        batch-size 1000
+        events (applications/get-dynamic-application-events-since from-id batch-size)
+        until-id (:event/id (last events))]
+    (if (empty? events)
+      state
+      (do
+        (log/info "Updating projection from" from-id "until" until-id)
+        (when (= batch-size (count events))
+          ;; there may be more events, so handle the next batch immediately after the current one
+          (trigger-applications-update!))
+        (assoc state
+               :last-processed-event-id until-id
+               :applications (reduce applications-view (:applications state) events))))))
+
+(defn trigger-applications-update! []
+  (send-off applications-state update-applications!))
+
+(comment
+  (send applications-state reset-applications-state)
+  (trigger-applications-update!)
+  @applications-state)
+
+(defstate applications-update-scheduler
+  :start (future (while (not (Thread/interrupted))
+                   (trigger-applications-update!)
+                   (Thread/sleep (-> 60 time/seconds time/in-millis))))
```

Comment:
As said in `applications-state` docs, the idea is that `trigger-applications-update!` will be called after new events have been created.

Reply:
There's also the option of checking for updates before each read operation, if extra consistency is needed.

Reply:
if the view is not always up to date, I think there'll be tricky to debug consistency related problems in our future. Wouldn't it be possible to always eagerly update the view so that every time you do something you can read a new up to date view that at least has what you updated? Eager updates could happen in the thread that causes them.

Reply:
Yes, that's also an option. Trigger the update after committing new events and block until all projections have been updated. Currently the commit happens in `rems.api/transaction-middleware`, so the update would need to be done after that.

When the system starts, it also needs to do the update (starting from blank state). Should that also block so that the application starts only after the projections are up to date?

Reply:
Yes at startup I think it would be best to block until the projection is up to date. Otherwise there can be surprising behavior. I guess most of our business logic happens with no need to use the updated projection so middleware is ok approach? But then keeping it synchronous makes no sense but it could happen eagerly.

Reply:
If we want to write business logic that uses the projection then we should keep the projection up to date always because otherwise there will just be unnecessarily hairy bugs IMHO.

Reply:
`trigger-applications-update!` is now called after commit.